lib/schema-controller.js

```diff
@@ -0,0 +1,169 @@
+
+/**
+ * When the dev will execute `fastify.validatorCompiler` this method will be called.
+ */
+SchemaController.prototype.getValidatorCompiler = function () {
+  return this.validatorCompiler || (this.parent && this.parent.getValidatorCompiler())
+}
+
+/**
+ * When the dev will execute `fastify.serializerCompiler` this method will be called.
+ */
+SchemaController.prototype.getSerializerCompiler = function () {
+  return this.serializerCompiler || (this.parent && this.parent.getSerializerCompiler())
+}
+
+/**
+ * This method will be called when a validator must be setup.
+ * Do not setup the compiler more than once
+ * @param {object} serverOptions: the fastify server option
+ */
+SchemaController.prototype.setupValidator = function (serverOption) {
+  const isReady = this.validatorCompiler !== undefined && !this.schemaBucket.hasNewSchemas()
+  if (isReady) {
+    return
+  }
+  this.validatorCompiler = buildPerformanceValidator(this.schemaBucket.getSchemas(), serverOption.ajv)
+}
+
+/**
+ * This method will be called when a serializer must be setup.
+ * Do not setup the compiler more than once
+ * @param {object} serverOptions: the fastify server option
+ */
+SchemaController.prototype.setupSerializer = function (serverOption) {
+  const isReady = this.serializerCompiler !== undefined && !this.schemaBucket.hasNewSchemas()
+  if (isReady) {
+    return
+  }
+  this.serializerCompiler = buildDefaultSerializer(this.schemaBucket.getSchemas())
+}
+
+module.exports = SchemaController
+
+function buildDefaultSerializer (externalSchemas) {
+  return function ({ schema, method, url, httpStatus }) {
+    return fastJsonStringify(schema, { schema: externalSchemas })
+  }
+}
+
+function ValidatorSelector () {
+  const validatorPool = new Map()
+  const cache = new Map()
+  cache.put = cache.set
+
+  return function buildCompilerFromPool (externalSchemas, options) {
+    const externals = JSON.stringify(externalSchemas)
+    const ajvConfig = JSON.stringify(options.customOptions)
+
+    const uniqueAjvKey = `${externals}${ajvConfig}`
+    if (validatorPool.has(uniqueAjvKey)) {
+      return validatorPool.get(uniqueAjvKey)
+    }
+
+    const compiler = ValidatorCompiler(externalSchemas, options, cache)
+    validatorPool.set(uniqueAjvKey, compiler)
+
+    return compiler
+  }
+}
+
+function ValidatorCompiler (externalSchemas, options, cache) {
+  // This instance of Ajv is private
+  // it should not be customized or used
+  const ajv = new Ajv(Object.assign({
+    coerceTypes: true,
+    useDefaults: true,
+    removeAdditional: true,
+    // Explicitly set allErrors to `false`.
+    // When set to `true`, a DoS attack is possible.
+    allErrors: false,
+    nullable: true
+  }, options.customOptions, { cache }))
+
+  if (options.plugins && options.plugins.length > 0) {
+    for (const plugin of options.plugins) {
+      plugin[0](ajv, plugin[1])
+    }
+  }
+
+  Object.values(externalSchemas).forEach(s => ajv.addSchema(s))
```

Comment:
Can you use for..of instead?

Reply:
FWIW, here is a full benchmark of different ways to do looping in Node:
| Node    | Option        | Msecs/op       | Ops/sec     | V8                    |
| ------- | ------------- | -------------- | ----------- | --------------------- |
| 15.5.0  | forloop       | 0.000169 msecs | 5909552.026 | V8 8.6.395.17-node.23 |
| 12.20.0 | forloop       | 0.000174 msecs | 5758159.789 | V8 7.8.279.23-node.45 |
| 15.5.0  | forloopcached | 0.000176 msecs | 5696850.011 | V8 8.6.395.17-node.23 |
| 12.20.0 | forloopcached | 0.000178 msecs | 5614266.195 | V8 7.8.279.23-node.45 |
| 10.23.0 | forloop       | 0.000201 msecs | 4979180.747 | V8 6.8.275.32-node.59 |
| 10.23.0 | forloopcached | 0.000201 msecs | 4972760.666 | V8 6.8.275.32-node.59 |
| 14.15.3 | forloopcached | 0.000204 msecs | 4913944.283 | V8 8.4.371.19-node.17 |
| 14.15.3 | forloop       | 0.000204 msecs | 4913940.057 | V8 8.4.371.19-node.17 |
| 15.5.0  | foreach       | 0.000225 msecs | 4443645.609 | V8 8.6.395.17-node.23 |
| 12.20.0 | foreach       | 0.000235 msecs | 4260004.435 | V8 7.8.279.23-node.45 |
| 10.23.0 | forof         | 0.000254 msecs | 3931537.464 | V8 6.8.275.32-node.59 |
| 14.15.3 | foreach       | 0.000266 msecs | 3752696.958 | V8 8.4.371.19-node.17 |
| 15.5.0  | forof         | 0.000278 msecs | 3590826.759 | V8 8.6.395.17-node.23 |
| 12.20.0 | forof         | 0.000282 msecs | 3551835.889 | V8 7.8.279.23-node.45 |
| 14.15.3 | forof         | 0.000306 msecs | 3269819.181 | V8 8.4.371.19-node.17 |
| 10.23.0 | foreach       | 0.000567 msecs | 1764332.007 | V8 6.8.275.32-node.59 |
| 15.5.0  | forin         | 0.004549 msecs | 219843.502  | V8 8.6.395.17-node.23 |
| 14.15.3 | forin         | 0.004553 msecs | 219643.363  | V8 8.4.371.19-node.17 |
| 12.20.0 | forin         | 0.005209 msecs | 191958.147  | V8 7.8.279.23-node.45 |
| 10.23.0 | forin         | 0.006276 msecs | 159334.217  | V8 6.8.275.32-node.59 |

You can see implementations of each variant here: https://github.com/kibertoad/nodejs-benchmark-tournament/tree/master/loops/contestants

But tl;dr version is that `for (let i = 0; i < values.length; i++) {` is still the fastest option.

Reply:
That said, use-case at hand is somewhat different, because there is also `Object.values()` involved, not sure which of the options is faster. I assume you are not actually concerned with performance (this is a one-time operation, right?) but merely want to simplify the syntax?

Reply:
I'd rather use `for...of` to be consistent with the codebase.